### PR TITLE
Elizabeth/scale tuples

### PR DESCRIPTION
**Pull request by noot**
## Changes
Added encoding and decoding of tuples (ie. structs).  Also, in DecodeByteArray and DecodeInteger, I now return another parameter `bytesDecoded` which is the number of the input bytes that were decoded.  I need this to keep track of how many bytes we've decoded so far in DecodeTuple, since an encoded tuple is the concatenation of all the fields, encoded, inside the struct.   As well as adding more comments.  

## Tests:
```
go test ./codec/...
```

### CODE REVIEW

File: codec/decode_test.go [hunk 1]
```diff
@@ -2,75 +2,136 @@ package codec
...
+	{val: []byte{0x04, 0x01}, output: []byte{0x01}, bytesDecoded: 2},
+	{val: []byte{0x04, 0xff}, output: []byte{0xff}, bytesDecoded: 2},
+	{val: []byte{0x08, 0x01, 0x01}, output: []byte{0x01, 0x01}, bytesDecoded: 3},
+	{val: append([]byte{0x01, 0x01}, byteArray(64)...), output: byteArray(64), bytesDecoded: 66},
+	{val: append([]byte{0xfd, 0xff}, byteArray(16383)...), output: byteArray(16383), bytesDecoded: 16385},
+	{val: append([]byte{0x02, 0x00, 0x01, 0x00}, byteArray(16384)...), output: byteArray(16384), bytesDecoded: 16388},
+	{val: append([]byte{0xfe, 0xff, 0xff, 0xff}, byteArray(1073741823)...), output: byteArray(1073741823), bytesDecoded: 1073741827},
 	// Causes CI to crash
-	//{val: append([]byte{0x03, 0x00, 0x00, 0x00, 0x40}, byteArray(1073741824)...), output: byteArray(1073741824)},
+	{val: append([]byte{0x03, 0x00, 0x00, 0x00, 0x40}, byteArray(1073741824)...), output: byteArray(1073741824), bytesDecoded: 1073741829},
```
Comment (ansermino) [inline codec/decode_test.go]: Comment out

File: codec/decode.go [hunk 2]
```diff
@@ -3,86 +3,101 @@ package codec
...
 
 // DecodeByteArray accepts a byte array representing a SCALE encoded byte array and performs SCALE decoding
-// of the byte array, then returns it.  If it is invalid, return nil and error
-func DecodeByteArray(b []byte) ([]byte, error) {
-	var o []byte
-	var err error
+// of the byte array
+// if the encoding is valid, it then returns (o, bytesDecoded, err) where o is the decoded byte array, bytesDecoded is
+// the number of input bytes decoded, and err is nil
+// otherwise, it returns nil, 0, and error
```
Comment (ansermino) [inline codec/decode.go]: please simplify comment. Most of this should be implied by function signature, generally godocs don't have explicit return details like this

Comment (ansermino) [reply codec/decode.go]: Should also explicity state if bytesDecoded includes the mode and length or not

File: codec/decode.go [hunk 3]
```diff
@@ -111,3 +127,64 @@ func DecodeBool(b byte) (bool, error) {
 
 	return false, errors.New("cannot decode invalid boolean")
 }
+
+// DecodeTuple acceps a byte array representing the SCALE encoded tuple and an interface. This interface should be a pointer
+// to a struct which the encoded tuple shpuld be marshalled into. If it is a valid encoding for the struct, it returns the
+// decoded struct and nil, otherwise return nil, err.
+// Note that we return the same interface that was passed to this function; this is because we are writing directly to the
+// struct that is passed in, using reflect to get each of the fields.
+func DecodeTuple(b []byte, t interface{}) (interface{}, error) {
```
Comment (ansermino) [inline codec/decode.go]: Same as above, the function signature already says some of this

File: codec/decode.go [hunk 4]
```diff
@@ -111,3 +127,64 @@ func DecodeBool(b byte) (bool, error) {
 
 	return false, errors.New("cannot decode invalid boolean")
 }
+
+// DecodeTuple acceps a byte array representing the SCALE encoded tuple and an interface. This interface should be a pointer
```
Comment (dutterbutter) [inline codec/decode.go]: accepts*

File: codec/decode.go [hunk 5]
```diff
@@ -111,3 +127,64 @@ func DecodeBool(b byte) (bool, error) {
...
+// to a struct which the encoded tuple shpuld be marshalled into. If it is a valid encoding for the struct, it returns the
```
Comment (dutterbutter) [inline codec/decode.go]: should*

File: codec/decode.go (hunk 2 above)
Comment (noot) [reply codec/decode.go]: comment simplified

File: codec/encode.go [hunk 6]
```diff
@@ -17,8 +18,12 @@ func Encode(b interface{}) ([]byte, error) {
 		return encodeInteger(int(v))
 	case int64:
 		return encodeInteger(int(v))
+	case string:
+		return encodeByteArray([]byte(v))
```
Comment (ansermino) [inline codec/encode.go]: Does this mean string encoding is complete?

File: codec/decode.go [hunk 7]
```diff
@@ -3,86 +3,101 @@ package codec
...
 
-// DecodeInteger accepts a byte array representing a SCALE encoded integer and performs SCALE decoding of the
-// int, then returns it
-func DecodeInteger(b []byte) (int64, error) {
-	if len(b) == 1 {
-		return int64(b[0] >> 2), nil
-	} else if len(b) == 2 {
+// DecodeInteger accepts a byte array representing a SCALE encoded integer and performs SCALE decoding of the int
+// if the encoding is valid, it then returns (o, bytesDecoded, err) where o is the decoded integer, bytesDecoded is the
+// number of input bytes decoded, and err is nil
```
Comment (ansermino) [inline codec/decode.go]: still confused about what bytesDecoded is. number of bytes is the length of the decoded integer?

File: codec/encode.go (hunk 6 above)
Comment (noot) [reply codec/encode.go]: yes, this should work for strings. needs tests though

File: codec/decode.go (hunk 7 above)
Comment (noot) [reply codec/decode.go]: bytesDecoded is the total number of bytes that were used to create the decoded integer so this includes the mode/length byte

**Pull request merged**